src/framework/components/element/pointer-events.js

```diff
@@ -0,0 +1,190 @@
+            for(var i = 0; i < this.entity.children.length; i++) {
+                var element =  this.entity.children[i];
+                if (element && element.element) {
+                    element.element[ name ].apply( element.element, eventData );
+                }
+            }
+        },
+
+        // Handles "down" pointer event – might be coming from touch or
+        // a mouse.
+        _pointerEventDown: function(point) {
+            point = this._parentPointToLocalPoint(point);
+
+            if (!this._testPointerEvent(point)) {
+                return;
+            }
+
+            this._passPointerEventToChildren("_pointerEventDown", [ point ]);
+            this.fire(pc.POINTEREVENT_DOWN, point);
+        },
+
+        // Handles "up" pointer event – might be coming from touch or
+        // a mouse.
+        _pointerEventUp: function(point) {
+            point = this._parentPointToLocalPoint(point);
+
+            if (!this._testPointerEvent(point)) {
+                return;
+            }
+
+            this._passPointerEventToChildren("_pointerEventUp", [ point ]);
+            this.fire(pc.POINTEREVENT_CLICK, point);
+            this.fire(pc.POINTEREVENT_UP, point);
+        },
+
+        // Fires pointer leave event and also makes all children do so.
+        _ensurePointerLeaveEvent: function(point) {
+            this._pointerOver = false;
+            this.fire(pc.POINTEREVENT_LEAVE, point); 
+
+            this._passPointerEventToChildren( "_ensurePointerLeaveEvent", [ point ]);
+        },
+
+        // Handles "move" pointer event – might be coming from touch or
+        // a mouse.
+        _pointerEventMove: function(point) {
+            point = this._parentPointToLocalPoint(point);
+
+            if (!this._testPointerEvent(point)) {
+                if (this._pointerOver) {
+                    this._ensurePointerLeaveEvent(point);            
+                }
+
+                return;
+            }
+
+            this._passPointerEventToChildren("_pointerEventMove", [ point ]);
+
+            if (!this._pointerOver) {
+                this._pointerOver = true;
+                this.fire(pc.POINTEREVENT_ENTER, point);                
+            }
+
+            this.fire(pc.POINTEREVENT_MOVE, point);
+        },
+
+        // Handles "scroll" pointer event – might be coming from touch or
+        // a mouse.
+        _pointerEventScroll: function(point, amount) {
+            point = this._parentPointToLocalPoint(point);
+
+            if (!this._testPointerEvent(point)) {
+                return;
+            }
+
+            this._passPointerEventToChildren("_pointerEventScroll", [ point, amount ]);
+            this.fire(pc.POINTEREVENT_SCROLL, point, amount);
+        },
+
+        // Mouse-specific event handler.
+        _onMouseDown: function(mouseEvent) {
+            this._pointerEventDown( new pc.Vec3( mouseEvent.x, mouseEvent.y, 0 ) );
```

Comment:
What if developer need to figure out coordinate of event within space of panel on which button was clicked?
I think there is a bit more fundamental functionality is required here, so it allows developers to easier code their stuff, like getting bottom-left corner of element relative to bottom-left of screen (0, 0) - is kind of fundamental block that is useful and universally simple to be used in very different cases.
Doing specific implementation of something, might cover only some % of use cases, and if it is something like half of people, but other half left without a solution and this wont solve their problems, then we need to challenge the way we implement it in first place, and even if we do at all.